src/Microsoft.OData.Core/ODataBatchWriter.cs

```diff
@@ -422,473 +388,307 @@ void IODataBatchOperationListener.BatchOperationContentStreamDisposed()
+        /// <summary>
+        /// Starts a new changeset.
+        /// </summary>
+        protected abstract void WriteStartChangesetImplementation();
 
-            // Reset the cache of content IDs here. As per spec, content IDs are only unique inside a change set.
-            this.payloadUriConverter.Reset();
-            this.currentOperationContentId = null;
-        }
+        /// <summary>
+        /// Ends an active changeset.
+        /// </summary>
+        protected abstract void WriteEndChangesetImplementation();
 
         /// <summary>
-        /// Verifies that calling CreateOperationRequestMessage if valid.
+        /// Creates an <see cref="ODataBatchOperationResponseMessage"/> for writing an operation of a batch response.
         /// </summary>
-        /// <param name="synchronousCall">true if the call is to be synchronous; false otherwise.</param>
-        /// <param name="method">The Http method to be used for this request operation.</param>
-        /// <param name="uri">The Uri to be used for this request operation.</param>
         /// <param name="contentId">The Content-ID value to write in ChangeSet head.</param>
-        private void VerifyCanCreateOperationRequestMessage(bool synchronousCall, string method, Uri uri, string contentId)
-        {
-            this.ValidateWriterReady();
-            this.VerifyCallAllowed(synchronousCall);
-
-            if (this.rawOutputContext.WritingResponse)
-            {
-                this.ThrowODataException(Strings.ODataBatchWriter_CannotCreateRequestOperationWhenWritingResponse);
-            }
+        /// <returns>The message that can be used to write the response operation.</returns>
+        protected abstract ODataBatchOperationResponseMessage CreateOperationResponseMessageImplementation(string contentId);
 
-            ExceptionUtils.CheckArgumentStringNotNullOrEmpty(method, "method");
-            if (this.changeSetBoundary != null)
-            {
-                if (HttpUtils.IsQueryMethod(method))
-                {
-                    this.ThrowODataException(Strings.ODataBatch_InvalidHttpMethodForChangeSetRequest(method));
-                }
+        /// <summary>
+        /// Writes all the pending headers and prepares the writer to write a content of the operation.
+        /// </summary>
+        protected abstract void StartBatchOperationContent();
 
-                if (string.IsNullOrEmpty(contentId))
-                {
-                    this.ThrowODataException(Strings.ODataBatchOperationHeaderDictionary_KeyNotFound(ODataConstants.ContentIdHeader));
-                }
-            }
+        /// <summary>
+        /// Disposes the batch writer and set the 'OperationStreamRequested' batch writer state;
+        /// called after the flush operation(s) have completed.
+        /// </summary>
+        protected abstract void DisposeBatchWriterAndSetContentStreamRequestedState();

```

Comment:
>DisposeBatchWriterAndSetContentStreamRequestedState [](start = 32, length = 51)

Why is this defined in the base class? It's only called from the derived class -- just make it a private method of the ODataMultipartMixedBatchWriter.

Reply:
done.

---
In reply to: [140937404](https://github.com/OData/odata.net/pull/960#discussion_r140937404) [](ancestors = 140937404)